plat/common/bootinfo.c

```diff
@@ -113,3 +119,142 @@ void ukplat_bootinfo_print(void)
+	 */
+	prop_len = 0;
+	prop_min_len = (int)sizeof(fdt32_t) * (naddr + nsz);
+	regs = fdt_getprop(fdtp, nmem, "reg", &prop_len);
+	if (unlikely(!regs || prop_len < prop_min_len))
+		ukplat_bootinfo_crash("Bad 'reg' property or more than one memory bank.");
+
+	mem_sz = fdt64_to_cpu(regs[1]);
+	mem_base = fdt64_to_cpu(regs[0]);
+	if (unlikely(!RANGE_CONTAIN(mem_base, mem_sz,
+				    __BASE_ADDR, __END - __BASE_ADDR)))
+		ukplat_bootinfo_crash("Image outside of RAM");
+
+	mrd.vbase = (__vaddr_t)mem_base;
+	mrd.pbase = (__paddr_t)mem_base;
+	mrd.len   = __BASE_ADDR - mem_base;
+	mrd.type  = UKPLAT_MEMRT_FREE;
+	mrd.flags = UKPLAT_MEMRF_READ | UKPLAT_MEMRF_WRITE;
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, &mrd);
+	if (unlikely(rc < 0))
+		ukplat_bootinfo_crash("Could not add free memory descriptor");
+
+	mrd.vbase = (__vaddr_t)__END;
+	mrd.pbase = (__paddr_t)__END;
+	mrd.len   = mem_base + mem_sz - __END;
+	mrd.type  = UKPLAT_MEMRT_FREE;
+	mrd.flags = UKPLAT_MEMRF_READ | UKPLAT_MEMRF_WRITE;
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, &mrd);
+	if (unlikely(rc < 0))
+		ukplat_bootinfo_crash("Could not add free memory descriptor");
+}
+
+static void fdt_bootinfo_cmdl_mrd(struct ukplat_bootinfo *bi, void *fdtp)
+{
+	const void *fdt_cmdl;
+	int fdt_cmdl_len;
+	__sz cmdl_len;
+	int nchosen;
+	char *cmdl;
+
+	nchosen = fdt_path_offset(fdtp, "/chosen");
+	if (unlikely(!nchosen))
+		return;
+
+	fdt_cmdl = fdt_getprop(fdtp, nchosen, "bootargs", &fdt_cmdl_len);
+	if (unlikely(!fdt_cmdl || fdt_cmdl_len <= 0))
+		return;
+
+	cmdl = ukplat_memregion_alloc(fdt_cmdl_len + sizeof(CONFIG_UK_NAME) + 1,
+				      UKPLAT_MEMRT_CMDLINE,
+				      UKPLAT_MEMRF_READ |
+				      UKPLAT_MEMRF_MAP);
+	if (unlikely(!cmdl))
+		ukplat_bootinfo_crash("Command-line alloc failed\n");
+
+	cmdl_len = sizeof(CONFIG_UK_NAME);
+	strncpy(cmdl, CONFIG_UK_NAME, cmdl_len);
+	cmdl[cmdl_len - 1] = ' ';
+	strncpy(cmdl + cmdl_len, fdt_cmdl, fdt_cmdl_len);
+	cmdl_len += fdt_cmdl_len;
+	cmdl[cmdl_len] = '\0';
+
+	bi->cmdline = (__u64)cmdl;
+	bi->cmdline_len = (__u64)cmdl_len;
+}
+
+static void fdt_bootinfo_fdt_mrd(struct ukplat_bootinfo *bi, void *fdtp)
+{
+	struct ukplat_memregion_desc mrd = {0};
+	int rc;
+
+	rc = fdt_check_header(fdtp);
+	if (unlikely(rc))
+		ukplat_bootinfo_crash("Invalid DTB");
+
+	mrd.vbase = (__vaddr_t)fdtp;
+	mrd.pbase = (__paddr_t)fdtp;
+	mrd.len   = fdt_totalsize(fdtp);
+	mrd.type  = UKPLAT_MEMRT_DEVICETREE;
+	mrd.flags = UKPLAT_MEMRF_READ | UKPLAT_MEMRF_MAP;
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, &mrd);
+	if (unlikely(rc < 0))
+		ukplat_bootinfo_crash("Could not insert DT memory descriptor");
+}
+
+void ukplat_bootinfo_fdt_setup(void *fdtp)
+{
+	struct ukplat_bootinfo *bi;
+
+	bi = ukplat_bootinfo_get();
+	if (unlikely(!bi))
+		ukplat_bootinfo_crash("Invalid bootinfo");
+
+	fdt_bootinfo_fdt_mrd(bi, fdtp);
+	fdt_bootinfo_mem_mrd(bi, fdtp);
+	ukplat_memregion_list_coalesce(&bi->mrds);
```

Comment:
I just remembered why I did this. `fdt_bootinfo_cmdl_mrd` makes use of `ukplat_memregion_alloc` and, therefore, if we were to call `ukplat_memregion_list_coalesce` after it, it would imply that the list has already been coalesced/sorted before doing that allocation. So I will keep it like this and add a clarifying comment.